### ISSUE TITLE
Overlay/Dialog with custom OverlayContainer

### ISSUE THREAD

**Issue by reppners** [angular/components#4612]
#### Bug, feature request, or proposal:
Feature request / proposal

#### What is the use-case or motivation for changing an existing behavior?
Put an overlay over a component where the backdrop would only cover a custom OverlayContainer element.

Basically I was trying to get the functionality Panel from angular material 1:
https://material.angularjs.org/1.1.4/demo/panel

Looking at the source of Overlay seems to be the basic building block which already provides much of the functionality needed I saw that the container where the Overlay is appended to the DOM is provided as an injectable.

Overlay:
https://github.com/angular/material2/blob/master/src/lib/core/overlay/overlay.ts

OverlayContainer:
https://github.com/angular/material2/blob/master/src/lib/core/overlay/overlay-container.ts

Are there already any plans or guidelines on how this kind of architecture can/should be used to provide a custom OverlayContainer?

**Comment by jelbourn** [comment]
@reppners does providing a custom `OverlayContainer` not give you what you need to do this?

**Comment by reppners** [comment]
Here is a plunkr
http://plnkr.co/edit/859v6jFIcFoHGnjVwE0u?p=preview

There is a component as dialog content and a custom overlay container which is provided as `OverlayContainer`. I expected to see the console log when the custom overlay container is instantiated but it does not appear.

Is the way of overriding the `OverlayContainer` in `main.ts` wrong?

**Comment by reppners** [comment]
http://plnkr.co/edit/859v6jFIcFoHGnjVwE0u?p=preview

Got it to work in a basic way. There is a css hack involved to make the overlay's `position: fixed` relative to a nested DOM element. The custom parent element needs `-webkit-transform: translateZ(0);` set.

However in my real application I'm trying to overload the provider in some nested component in the providers section of the Component decorator. The component is inside a lazy loaded module. In that context it does not even run the constructor of the custom OverlayContainer.

To summarize taking the approach of using the hierarchical DI seems to work for this simple use-case.
What about a nested lazy-loaded component that wants to show some overlay appended to an element from its template for e.g. a card details view but also wants to show top-level dialogs as alerts? Would an indirection over a service defined in a core module use the original `OverlayContainer` when opening a dialog?

**Comment by jelbourn** [comment]
@reppners are you able to get what you need by creating your own custom OverlayModule? E.g.
```ts
@NgModule({
    providers: [
    Overlay,
    OverlayPositionBuilder,
    VIEWPORT_RULER_PROVIDER,
    PizzaOverlayContainer,
  ]
})
export class PizzaOverlayModule { }
```

And then using that custom module in place of overlay module everywhere?

**Comment by reppners** [comment]
I've published an implementation and a little demo of the functionality.

https://github.com/reppners/ngx-cdk-dynamic-overlay-container

Maybe it can help other people that need similar functionality.

**Comment by AsafAgranat** [comment]
@reppners 
Thanks for sharing your implementation Stefan - thorough and useful. 
Do you know perhaps, since you create the overlay dynamically, if it's possible to tap onto the dialog config and define a start position (`top `and `left ` in `px`) for the dialog - so it would animate-in from the trigger element (button)? 
This functionality was provided in Material v1, but is missing from Material +v2.

**Comment by reppners** [comment]
@AsafAgranat Unfortunately I don't have a solution at hand and also don't have the time to explore the possibilities to find one.

**Comment by dsadfasd** [comment]
> I've published an implementation and a little demo of the functionality.
> 
> https://github.com/reppners/ngx-cdk-dynamic-overlay-container
> 
> Maybe it can help other people that need similar functionality.

Hey... this example is not working angular 7.

**Comment by crisbeto** [comment]
Duplicate of https://github.com/angular/components/issues/3221.
